### PR TITLE
WIP Deprecate / discourage LCM threads

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
In modern C++, there is little need for explicit `std::thread` at all, and `std::mutex` is a dirty word.  (Basically nobody can write correct, mutex-using code.)  Developers should prefer higher-level primitives instead.

Furthermore, LCM itself has a background thread already!  Messages are still serviced from the kernel even if our application never launches our own thread, so we don't need `DrakeLcm` to manage a thread anyway.  If the user wants async wakeups racing their data, then can write that application-level code on their own.  Most uses in Drake do _not_ want that -- indeed, the `LcmSubscriberSystem` goes to great lengths to _prevent_ new messages from being reflected until the `Simulator` tells it to change its outputs.

Another good reason to nix threads is pydrake + GIL problems (again, deadlocks and races).  If the python user does their own thread management entirely pythonically, they can suit it to meet their needs (or "keep the pieces").

There is also the idea that Drake is an optimization / dynamics / simulation / control theory toolbox, and should innovate robot middleware as little as possible.  Other libraries for application coordination / messaging do a great job, and we should not try to (badly) re-implement them.

So, bottom line, we can and should remove a bunch of dangerous LCM and thread-related code from Drake-the-library.  This is rough draft of what that might look like.  I will be carving off parts into PRs for review, but I also think it's important to be able to view (and test!) the final product prototyped here.

- New drake/lcm APIs
  - [x] #11041 
    - [x] #10926
    - [x] #11036
  - [x] #11175
- New drake/systems/lcm APIs
  - [x] #10935
  - [x] #10970
  - [x] #11121
     - [x] #11122
        - [x] #11140
- Porting
  - [x] #11124
  - [x] #11119
- Deprecation
  - [x] #11166
  - [x] #11127
  - [x] #11131

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/10981)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
As of the most recent push, the most acute missing piece in the prototype is how to replace (perhaps just with documentation) the LcmDrivenLoop.  It shouldn't need a thread -- it should just manually tick the diagram + tock the lcm queue, should use port.FixInput to inject the clocked LCM inputs into the simulation (not LcmSubscriberSystem), and probably should even publish the clocked outputs explicitly, instead as a side-effect of a simulation event.  I am probably going to focus on getting the DrakeLcmInterface and DrakeLcm improvements into PR, before pushing on the LcmDrivenLoop prototype.
